File: FixtureGenerators/Pooled.py
from FixtureGenerators.FixturesGenerator import FixturesGenerator
from database import db
from database.models import TournamentTeams, Tournaments, Games, Teams
from structure import manage_game


class Pooled(FixturesGenerator):

    def __init__(self, tournament_id):
        super().__init__(tournament_id, fill_officials=True, editable=False, fill_courts=True)

    def _begin_tournament(self, tournament_id):
        tournament = Tournaments.query.filter(Tournaments.id == tournament_id).first()
        teams = TournamentTeams.query.filter(TournamentTeams.tournament_id == tournament_id).all()
        teams = sorted(teams, key=lambda x: (x.team.elo() != 1500, -x.team.elo()))
        pool = 0
        for i in teams:
            i.pool = 1 + pool
            pool = 1 - pool
        tournament.is_pooled = True
        db.session.commit()

    def _end_of_round(self, tournament_id):
        tournament = Tournaments.query.filter(Tournaments.id == tournament_id).first()
        teams = TournamentTeams.query.filter(TournamentTeams.tournament_id == tournament_id).all()
        game = Games.query.filter(Games.tournament_id == tournament_id).order_by(Games.round.desc()).first()
        rounds = game.round if game else 0

        pools = [[j.team for j in teams if j.pool == i] for i in range(1, 3)]

        for pool in pools:
            if len(pool) % 2 != 0:
                pool += [Teams.BYE]

        if max(len(i) for i in pools) <= rounds + 1:
            tournament.in_finals = True
            return
        for _ in range(rounds):
            for pool in pools:
                pool[1:] = [pool[-1]] + pool[1:-1]
            # Rotate the teams except the first one
        for pool in pools:
            mid = len(pool) // 2
            for j in range(mid):
                team_one = pool[j]
                team_two = pool[len(pool) - 1 - j]
                manage_game.create_game(tournament_id, team_one.searchable_name, team_two.searchable_name,
                                        round_number=rounds + 1)

File: HandballBackend/FixtureGenerator/Pooled.cs
using HandballBackend.Database;
using HandballBackend.Database.Models;
using HandballBackend.EndpointHelpers.GameManagement;
using Microsoft.EntityFrameworkCore;

namespace HandballBackend.FixtureGenerator;

public class Pooled : AbstractFixtureGenerator {
    private readonly int _tournamentId;
    private readonly bool _blitz;


    public Pooled(int tournamentId, bool blitz = false) : base(tournamentId, true, true) {
        _tournamentId = tournamentId;
        _blitz = blitz;
    }

    public override async Task BeginTournament() {
        var db = new HandballContext();
        var tournament = (await db.Tournaments.FindAsync(_tournamentId))!;
        tournament.IsPooled = true;

        var teams = (await db.TournamentTeams
            .Where(t => t.TournamentId == _tournamentId)
            .IncludeRelevant()
            .ToArrayAsync()).OrderByDescending(t => t.Team.TrueElo()).ToList();

        var pool = 0;
        foreach (var team in teams) {
            team.Pool = 1 + pool;
            pool = 1 - pool;
        }

        await db.SaveChangesAsync();
        await base.BeginTournament();
    }

    public override async Task<bool> EndOfRound() {
        var db = new HandballContext();
        var tournament = (await db.Tournaments.FindAsync(_tournamentId))!;
        var tournamentTeams = await db.TournamentTeams
            .Where(t => t.TournamentId == _tournamentId)
            .IncludeRelevant()
            .ToListAsync();

        var rounds = await db.Games
            .Where(g => g.TournamentId == _tournamentId)
            .OrderByDescending(g => g.Round).Select(g => g.Round).FirstOrDefaultAsync();

        var poolOne = tournamentTeams.Where(tt => tt.Pool == 1).Select(tt => tt.Team).OrderBy(t => t.Id).ToList();
        var poolTwo = tournamentTeams.Where(tt => tt.Pool == 2).Select(tt => tt.Team).OrderBy(t => t.Id).ToList();
        if (poolOne.Count % 2 != 0) {
            poolOne.Add(db.Teams.First(t => t.Id == 1));
        }

        if (poolTwo.Count % 2 != 0) {
            poolTwo.Add(db.Teams.First(t => t.Id == 1));
        }

        if (Math.Min(poolTwo.Count, poolOne.Count) <= rounds + 1) {
            // we are now in finals
            tournament.InFinals = true;
            await db.SaveChangesAsync();
            return true;
        }

        for (var i = 0; i < rounds; i++) {
            poolOne.Insert(1, poolOne.Last());
            poolOne.RemoveAt(poolOne.Count - 1);
            poolTwo.Insert(1, poolTwo.Last());
            poolTwo.RemoveAt(poolTwo.Count - 1);
        }

        for (var i = 0; i < poolOne.Count / 2; i++) {
            var teamOne = poolOne[i];
            var teamTwo = poolOne[poolOne.Count - i - 1];
            await GameManager.CreateGame(_tournamentId, teamOne.Id, teamTwo.Id, blitzGame: _blitz,
                round: rounds + 1);
        }

        for (var i = 0; i < poolTwo.Count / 2; i++) {
            var teamOne = poolTwo[i];
            var teamTwo = poolTwo[poolTwo.Count - i - 1];
            await GameManager.CreateGame(_tournamentId, teamOne.Id, teamTwo.Id, blitzGame: _blitz,
                round: rounds + 1);
        }

        await db.SaveChangesAsync();
        return await base.EndOfRound();
    }
}